### DISCUSSION TITLE
Camel Fault Tolerance from MicroProfile not working in Native mode

### DISCUSSION THREAD

**Discussion by MikaelAnderssonWigander** [quarkusio/quarkus#43072] [Q&A]
Hi

In my Camel Quarkus application I'd like to use Camels Circuit breaker EIP together with their implementation of Fault Tolerance from MicroProfile.

https://camel.apache.org/components/4.4.x/eips/fault-tolerance-eip.html

However this extension is not working when compiling to native mode.

```
Error: Discovered unresolved invoke dynamic during parsing: org.apache.camel.component.microprofile.faulttolerance.FaultToleranceProcessor.process(org.apache.camel.Exchange, org.apache.camel.AsyncCallback). This error is reported at image build time because class org.apache.camel.component.microprofile.faulttolerance.FaultToleranceProcessor is registered for linking at image build time by command line and command line.
Error encountered while parsing org.apache.camel.component.microprofile.faulttolerance.FaultToleranceProcessor.process(FaultToleranceProcessor.java:253)
Parsing context:
   at org.apache.camel.processor.FinallyProcessor.process(FinallyProcessor.java:68)
   at org.apache.camel.impl.engine.CamelInternalProcessor.processNonTransacted(CamelInternalProcessor.java:346)
   at org.apache.camel.impl.engine.CamelInternalProcessor.process(CamelInternalProcessor.java:322)
   at org.apache.camel.processor.UnitOfWorkProducer.process(UnitOfWorkProducer.java:58)
   at org.apache.camel.processor.InterceptSendToEndpointProcessor.callback(InterceptSendToEndpointProcessor.java:114)
   at org.apache.camel.processor.InterceptSendToEndpointProcessor.process(InterceptSendToEndpointProcessor.java:83)
   at root method.(Unknown Source)

Detailed message:

com.oracle.svm.core.util.UserError$UserException: Discovered unresolved invoke dynamic during parsing: org.apache.camel.component.microprofile.faulttolerance.FaultToleranceProcessor.process(org.apache.camel.Exchange, org.apache.camel.AsyncCallback). This error is reported at image build time because class org.apache.camel.component.microprofile.faulttolerance.FaultToleranceProcessor is registered for linking at image build time by command line and command line.
```

Any ideas if this is something that can be fixed with `additional-build-args` or if it is a feature in Camels implementation?

Link to repo for ref:
https://github.com/MikaelAnderssonWigander/fault-tolerance-camel

#NOTE: no code needed, just build with -Pnative

If code is required, please let me know and I'll put it some but it will give the same result

Environment:

Java version: 22.0.1+8, vendor version: Oracle GraalVM 22.0.1+8.1
 Graal compiler: optimization level: 2, target machine: x86-64-v3, PGO: off
 C compiler: cc (apple, x86_64, 15.0.0)

Thx

[a comment by an automated account is not shown]

**Comment by Ladicek** [comment]
For the record, the GitHub repo https://github.com/MikaelAnderssonWigander/fault-tolerance-camel doesn't contain any source code.

I'm not sure what Camel does with SmallRye Fault Tolerance, and I'm not sure who would the proper contact person be. CC @ppalaga

**Reply by MikaelAnderssonWigander** [reply]
It's on purpose, just build -Pnative

**Comment by ppalaga** [comment]
> If code is required, please let me know and I'll put it some but it will give the same result

It would be nice! What would be the point of fixing something that does not do anything useful? Your reproducer can eventually be transformed to a test so a route that shows the behavior would be nice.

**Reply by MikaelAnderssonWigander** [reply]
I'll add some code then but that will not alter the outcome.

It can be done with just a build…

**Reply by ppalaga** [reply]
Thanks @MikaelAnderssonWigander, it seems to be [this problem](https://github.com/apache/camel-quarkus/issues/6285) related to misalignment of SmallRye Fault Tolerance versions between Camel and Quarkus. The problem should not occur with earlier versions of Quarkus Platform, such as 3.13.0

**Reply by MikaelAnderssonWigander** [reply]
The weird thing is that I have another repo using these dependencies and that one is working.

**Reply by MikaelAnderssonWigander** [reply]
Back again in 3.16.3, working in 3.15.1

**Comment by MikaelAnderssonWigander** [comment]
Code added… Same outcome…

**Comment by MikaelAnderssonWigander** [comment]
Added a new branch, ```working-version``` that's working.
The only difference is that this includes an Oracle dependency and a argument added to build;
``` --initialize-at-run-time=oracle.jdbc.driver.OracleDriver```

**Comment by MikaelAnderssonWigander** [comment]
Hi. This error has reappeared in 3.16.3. 
It works in 3.15.1

**Reply by ppalaga** [reply]
@MikaelAnderssonWigander I guess there is no change with Quarku Platform 3.17.0, because Camel Quarkus did not release 3.17.0 yet. We wait for Camel 3.9.0 that should solve this problem.
